### DISCUSSION TITLE
How scalable is Shadow?

### DISCUSSION THREAD

**Discussion by lifeoftomi** [shadow/shadow#1753] [General]
According to https://shadow.github.io/docs/guide/, Shadow can simulate 1000s of network nodes and run real applications in a scalable way.

As a reference point, I'm curious to know how the [shadow-plugin-bitcoin](https://github.com/shadow/shadow-plugin-bitcoin) worked and how many nodes it could run. Bitcoin needs specialized hardware and is computationally intensive, so how does that work in Shadow?

**Comment by robgjansen** [comment]
Shadow's [scalability](https://en.wikipedia.org/wiki/Scalability) is a characteristic that generally refers to how efficient Shadow is as the size of a simulation (the number of nodes and types of computation load they add) increases. I like to think of this more like _how much overhead does Shadow add on top of running the applications and how does the overhead change as the size of the simulation changes_.

To answer this question directly, Shadow itself (the code inside of the simulator) is very efficient. We are using the most efficient methods known for launching new applications, for intercepting system calls made by applications running in Shadow, and for transferring state between Shadow and the applications. We have some more thorough technical analysis in a research paper that we are not yet ready to share.

But I suspect what you really want to know is how many nodes can you run in Shadow. That depends _entirely_ on the applications that you run in Shadow, and Shadow does not impose any restrictions here. You are only limited by the resources available on the host machine---both memory and computational resources---and the time you are willing to wait to complete the simulation. Shadow doesn't really do any tricks to "cancel out" any computational load that would be performed by the application (although the Tor-specific plugin that we no longer support used to skip some cryptographic cipher operations (#1370)).

The simulation itself is composed of the sum of the computational tasks of the applications running in your simulation. For example, if you have an application that consumes 100 GB of RAM and continuously compresses very large amounts of data, you will likely want to run fewer instances of that application than if you were running a simple `ping` client that sends a small packet once every simulated minute. So yes, Shadow can run 1000s of nodes, and this has been done in large scale studies of the Tor network (see for example [a recently published Tor research paper](https://www.robgjansen.com/publications/neverenough-sec2021.pdf)). But it might take a long time to complete a 1000 node simulation depending on which applications you are running and how much CPU load your applications are performing. One benefit of using Shadow is that this work will be performed _correctly in simulation_ whereas you would almost certainly not be able to run 1000 nodes in real time on your machine without considerable slowdown/distortion.

For our [shadow-bitcoin](https://www.robgjansen.com/publications/shadowbitcoin-cset2015.pdf) paper, we were able to run 6000 nodes after taking some shortcuts that were specific to that research. That was using the old 1.x version of Shadow, and our design has changed significantly since then (for the better). See the paper for more specifics.
